In [2]:
import pandas as pd

df = pd.read_csv("../data/titanic.csv") # Load the titanic dataset

df.shape # Show the number of rows and columns of a dataset

(891, 12)

Showing the first few rows of the dataset

In [21]:
df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


In [24]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

In [28]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

In [26]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


In [29]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [30]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [3]:
(df.isnull().sum())/len(df)*100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

The dataset was checked for duplicate records using `df.duplicated().sum()`.

Result:
- No duplicate rows were found in the dataset.

Therefore, no duplicate removal was required.

In [ ]:
df.duplicated().sum()
# df = df.drop_duplicates()

np.int64(0)

In [7]:
missing = df.isnull().sum()
missing_percentage = (missing/len(df))*100

pd.DataFrame({
    "Missing values":missing,
    "Percentage":missing_percentage
}).sort_values("Missing values",ascending=False)

,Missing values,Percentage
Cabin,687,77.104377
Age,177,19.865320
Embarked,2,0.224467
PassengerId,0,0.000000
Name,0,0.000000
Pclass,0,0.000000
Survived,0,0.000000
Sex,0,0.000000
Parch,0,0.000000
SibSp,0,0.000000


In [8]:
df["Cabin"].isnull().sum()

np.int64(687)

## Missing Value Analysis

Three columns contain missing values:

| Column | Missing Values | Percentage |
|---------|---------|---------|
| Cabin | 687 | 77.10% |
| Age | 177 | 19.87% |
| Embarked | 2 | 0.22% |

Observations:

- Cabin contains a very large number of missing values and may be considered for removal.
- Age contains a moderate number of missing values and should be imputed.
- Embarked contains only two missing values and can be handled easily through imputation or row removal.

In [10]:
df[df["Cabin"].isnull()].head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S


In [11]:
df[df["Age"].isnull()].head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
17,18,1,2,"Williams, Mr. Charles Eugene",male,NaN,0,0,244373,13.0000,NaN,S
19,20,1,3,"Masselmani, Mrs. Fatima",female,NaN,0,0,2649,7.2250,NaN,C
26,27,0,3,"Emir, Mr. Farred Chehab",male,NaN,0,0,2631,7.2250,NaN,C
28,29,1,3,"O'Dwyer, Miss. Ellen ""Nellie""",female,NaN,0,0,330959,7.8792,NaN,Q


In [12]:
df[df["Embarked"].isnull()]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,1,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


Analyzing the above 2 rows further instead using simple imputation at first (Mode/Median imputation)

In [13]:
df[df["Cabin"]=="B28"]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,1,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


In [14]:
df[df["Ticket"]=="113572"]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,1,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


In [15]:
df["Embarked"].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [18]:
df.groupby("Embarked")["Fare"].mean()

Embarked
C    59.954144
Q    13.276030
S    27.243651
Name: Fare, dtype: float64

## Embarked Missing Value Analysis

The `Embarked` column contained 2 missing values.

Investigation showed that both passengers:

- Traveled in First Class (Pclass = 1)
- Had the same ticket number (113572)
- Occupied the same cabin (B28)
- Paid a fare of 80.0

Average fare analysis by embarkation port showed:

| Embarked | Average Fare |
|-----------|-------------:|
| C | 59.95 |
| Q | 13.28 |
| S | 27.24 |

Since the fare of 80.0 is most consistent with passengers embarking from Cherbourg (`C`), the missing values were imputed with `C`.

This approach was chosen instead of simple mode imputation because it uses contextual information from the dataset.

In [19]:
df.loc[df["Embarked"].isnull(),"Embarked"]="C"
df["Embarked"].isnull().sum()

np.int64(0)

## Cabin Feature Analysis and Treatment

The `Cabin` column was identified as having a significant amount of missing data.

### Missing Value Statistics

- Total Records: 891
- Missing Cabin Values: 687
- Missing Percentage: 77.10%

### Investigation

Analysis of the `Cabin` column revealed:

- A very high percentage of missing values.
- Numerous unique cabin identifiers such as `B28`, `C85`, `E46`, and `G6`.
- Most cabin values occurred only a few times.
- Direct imputation using mode, mean, or median was not appropriate because cabin numbers represent categorical location information rather than numerical measurements.

### Feature Engineering

Although the full cabin identifier contained too many missing values, the first character of the cabin value represents the passenger's deck.

Examples:

| Cabin | Deck |
|---------|---------|
| B28 | B |
| C85 | C |
| E46 | E |
| G6 | G |

To preserve potentially useful information, a new feature named `Deck` was created by extracting the first character from the `Cabin` column.

### Final Decision

After extracting the deck information:

- The original `Cabin` column was removed.
- The new `Deck` feature was retained for future analysis and modeling.

This approach preserves meaningful information while eliminating a feature with excessive missing values.

In [21]:
cabin_missing = df["Cabin"].isnull().sum()
cabin_miss_percentage = (cabin_missing/len(df))*100

print("Missing values :",cabin_missing)
print("Missing Percentage: ",cabin_miss_percentage)

Missing values : 687
Missing Percentage:  77.10437710437711


In [22]:
df["Cabin"].value_counts().head()

Cabin
G6             4
C23 C25 C27    4
B96 B98        4
F33            3
E101           3
Name: count, dtype: int64

In [23]:
df["Deck"]= df["Cabin"].str[0]
df["Deck"].value_counts(dropna=False)

Deck
NaN    687
C       59
B       47
D       33
E       32
A       15
F       13
G        4
T        1
Name: count, dtype: int64

In [24]:
df.drop(columns=["Cabin"],inplace=True)

## Age Feature Analysis and Missing Value Investigation

The `Age` column contains missing values and is an important feature for predicting passenger survival.

### Missing Value Statistics

- Total Records: 891
- Missing Age Values: 177
- Missing Percentage: 19.87%

### Importance of the Feature

Age is considered a significant variable because survival rates on the Titanic varied across different age groups.

Examples:

- Children were often prioritized during evacuation.
- Adults and elderly passengers may have experienced different survival outcomes.
- Age can influence travel patterns and passenger characteristics.

Therefore, removing the `Age` column would result in the loss of potentially valuable information.

### Investigation Strategy

Before selecting an imputation method, the distribution of the `Age` feature must be analyzed.

The following questions need to be answered:

1. Is the distribution approximately symmetric?
2. Is the distribution skewed?
3. Are there significant outliers?
4. Which measure of central tendency best represents the data?

### Imputation Decision Criteria

| Distribution Type | Preferred Imputation Method |
|-------------------|----------------------------|
| Approximately Normal | Mean |
| Skewed Distribution | Median |
| Categorical Data | Mode |

To determine the most appropriate imputation technique, the distribution and skewness of the `Age` feature will be examined using statistical measures and visualizations.

### Next Steps

The following analyses will be performed:

- Summary statistics of Age
- Distribution visualization
- Skewness calculation
- Outlier investigation

The results of these analyses will guide the selection of an appropriate imputation strategy.

In [25]:
df["Age"].skew()

np.float64(0.38910778230082704)

In [27]:
df.groupby("Pclass")["Age"].median()

Pclass
1    37.0
2    29.0
3    24.0
Name: Age, dtype: float64

In [28]:
df.groupby("Sex")["Age"].median()

Sex
female    27.0
male      29.0
Name: Age, dtype: float64

In [31]:
df["Age"] = df.groupby(["Pclass","Sex"])["Age"].transform(lambda x: x.fillna(x.median))

## Age Imputation Strategy Selection

The Age column contains 177 missing values (19.87% of the dataset).

### Distribution Analysis

- Skewness: 0.389
- The distribution is approximately symmetric.
- Mean imputation would normally be considered appropriate.

### Group Analysis

Median age by passenger class:

| Pclass | Median Age |
|---------|-----------|
| 1 | 37 |
| 2 | 29 |
| 3 | 24 |

The substantial difference between passenger classes indicates that age is associated with passenger class.

### Decision

Instead of using a single global mean or median for all passengers, a group-wise median imputation approach will be used.

This approach preserves age differences between passenger classes and produces more realistic estimates for missing values.

### Selected Method

Group-wise median imputation based on passenger class.

Benefits:

- Reduces imputation bias
- Preserves class-specific age characteristics
- More representative than global mean imputation